clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -698,36 +753,54 @@ synchronized boolean hasUnresolvedSequences() {
     }
 
     synchronized boolean hasUnresolvedSequence(TopicPartition topicPartition) {
-        return partitionsWithUnresolvedSequences.contains(topicPartition);
+        return partitionsWithUnresolvedSequences.containsKey(topicPartition);
     }
 
-    synchronized void markSequenceUnresolved(TopicPartition topicPartition) {
-        log.debug("Marking partition {} unresolved", topicPartition);
-        partitionsWithUnresolvedSequences.add(topicPartition);
+    synchronized void markSequenceUnresolved(ProducerBatch batch) {
+        int nextSequence = batch.baseSequence() + batch.recordCount;
```

Comment:
Would it be clearer to use `batch.lastSequence + 1`?

Reply:
`ProducerBatch` doesn't have a `lastSequence`. Are you suggesting moving this computation into the batch object instead of doing it here?

Reply:
I guess I was getting confused with `RecordBatch`. We could expose `lastSequence` from `MemoryRecordsBuilder` potentially.